### PR TITLE
Enable quantizer op on ROCm

### PR DESCRIPTION
**Pull request by rraminen**
This PR is to enable quantizer extension on ROCm

cc: @jithunnair-amd

### CODE REVIEW

**Comment by jithunnair-amd** [comment]
> /opt/rocm-5.3.2/

You might need to update your ROCm version. The latest release is ROCm5.6

File: op_builder/quantizer.py [hunk 1]
```diff
@@ -28,7 +28,14 @@ def sources(self):
         ]
 
     def include_paths(self):
-        return ['csrc/includes']
+        includes = ['csrc/includes']
+        if self.is_rocm_pytorch():
+            from torch.utils.cpp_extension import ROCM_HOME
+            includes += ['{}/hiprand/include'.format(ROCM_HOME), '{}/rocrand/include'.format(ROCM_HOME)]
```
[the remark this answers is not in the document]
Comment (jithunnair-amd) [reply op_builder/quantizer.py:34]: Also, https://github.com/pytorch/pytorch/blob/main/torch/utils/cpp_extension.py#L1162 already adds the `$ROCM_HOME/include` path for any PyTorch extensions.

As for the ROCm library headers, https://github.com/pytorch/pytorch/blob/fe594ab3235594dae807eb909af8ba597423ad37/torch/utils/hipify/cuda_to_hip_mappings.py#L611 is an example of what the hipify mapping would look like to work with the above include path. Can you please make sure any required header mappings for this extension are added to `cuda_to_hip_mappings.py`?

**Comment by rraminen** [comment]
> Also seeing this with 5.5 (the latest version I have an Azure container for):
> 
> ```
>       In file included from csrc/includes/quantization_utils_hip.h:12:
>       csrc/includes/reduction_utils_hip.h:278:43: error: excess elements in struct initializer
>           constexpr __half2_raw zero = {0x0000, 0x0000};
>                                                 ^~~~~~
>       csrc/includes/reduction_utils_hip.h:285:42: error: excess elements in struct initializer
>           constexpr __half2_raw inf = {0x7C00, 0x7C00};
>                                                ^~~~~~
>       csrc/includes/reduction_utils_hip.h:292:46: error: excess elements in struct initializer
>           constexpr __half2_raw neg_inf = {0xFC00, 0xFC00};
>                                                    ^~~~~~
>       csrc/includes/reduction_utils_hip.h:412:18: error: no member named 'meta_group_size' in 'cooperative_groups::thread_block_tile<64, void>'
>           if (warp_arg.meta_group_size() > 1 && total_warps != 1) {
>               ~~~~~~~~ ^
>       csrc/includes/reduction_utils_hip.h:417:54: error: no member named 'meta_group_rank' in 'cooperative_groups::thread_block_tile<64, void>'
>                           reduce_buffer + elems * warp_arg.meta_group_rank() + i, data + i);
>                                                   ~~~~~~~~ ^
>       csrc/includes/reduction_utils_hip.h:424:22: error: no member named 'meta_group_rank' in 'cooperative_groups::thread_block_tile<64, void>'
>               if (warp_arg.meta_group_rank() == 0) {
>                   ~~~~~~~~ ^
>       csrc/includes/reduction_utils_hip.h:425:51: error: no member named 'meta_group_size' in 'cooperative_groups::thread_block_tile<64, void>'
>                   if (warp_arg.thread_rank() < warp_arg.meta_group_size()) {
>                                                ~~~~~~~~ ^
>       csrc/includes/reduction_utils_hip.h:450:69: error: no member named 'meta_group_rank' in 'cooperative_groups::thread_block_tile<64, void>'
>                                                  reduce_buffer + warp_arg.meta_group_rank() * elems + i);
>                                                                  ~~~~~~~~ ^
>       csrc/includes/reduction_utils_hip.h:523:38: error: no member named 'meta_group_rank' in 'cooperative_groups::thread_block_tile<64, void>'
>               const int warp_offset = warp.meta_group_rank() & ~(num_warps - 1);
>                                       ~~~~ ^
>       csrc/includes/reduction_utils_hip.h:540:38: error: no member named 'meta_group_rank' in 'cooperative_groups::thread_block_tile<64, void>'
>               const int warp_offset = warp.meta_group_rank() & ~(num_warps - 1);
>                                       ~~~~ ^
>       csrc/includes/reduction_utils_hip.h:561:38: error: no member named 'meta_group_rank' in 'cooperative_groups::thread_block_tile<64, void>'
>               const int warp_offset = warp.meta_group_rank() & ~(num_warps - 1);
>                                       ~~~~ ^
>       csrc/includes/reduction_utils_hip.h:584:38: error: no member named 'meta_group_rank' in 'cooperative_groups::thread_block_tile<64, void>'
>               const int warp_offset = warp.meta_group_rank() & ~(num_warps - 1);
>                                       ~~~~ ^
>       12 errors generated when compiling for gfx908.
>       error: command '/opt/rocm-5.5.0/bin/hipcc' failed with exit code 1
>       [end of output]
> ```

Hi @loadams , the solution to these errors will be available from ROCm 6.0. So we might have to wait till then. Please let me know if it is a good idea to close this PR now and file it again when fixes are available in ROCm 6.0.

**Comment by pruthvistony** [comment]
> Hi @rraminen - its fine to leave this PR open and we can just re-run it/complete it when ROCm 6.0 is released. Do you know of an ETA for that?

@loadams ,
ROCm 6.0 is major version release and it is currently targeted  by 2023 year end roughly.

**Comment by rraminen** [comment]
Hi @loadams , the meta_group_size() and meta_group_rank() APIs are now available in cooperative_groups on ROCm. Could you please re-run it/complete this? Thanks.

**Comment by rraminen** [comment]
This requires rocm5.7 or later.

**Comment by loadams** [comment]
> This requires rocm5.7 or later.

Thanks, we will try to get our MI200 nodes running, currently have an infrastructure issue, but will run tests and then get this merged.

**Comment by loadams** [comment]
> > This requires rocm5.7 or later.
> 
> Thanks, we will try to get our MI200 nodes running, currently have an infrastructure issue, but will run tests and then get this merged.

@rraminen - one other question, this isn't backwards compatible, right?  So users would need to use ROCm > 5.7 if they're using from this commit forward, right?

**Comment by jithunnair-amd** [comment]
> > > This requires rocm5.7 or later.
> > 
> > 
> > Thanks, we will try to get our MI200 nodes running, currently have an infrastructure issue, but will run tests and then get this merged.
> 
> @rraminen - one other question, this isn't backwards compatible, right? So users would need to use ROCm > 5.7 if they're using from this commit forward, right?

@loadams Thanks for bringing up the backward-compatibility concern. This PR "per se" is backward-compatible, but the "quantizer" extension itself might not be supported pre-ROCm5.7 because it needs the changes to the HIP cooperative groups headers.

**Comment by loadams** [comment]
> > > > This requires rocm5.7 or later.
> > > 
> > > 
> > > Thanks, we will try to get our MI200 nodes running, currently have an infrastructure issue, but will run tests and then get this merged.
> > 
> > 
> > @rraminen - one other question, this isn't backwards compatible, right? So users would need to use ROCm > 5.7 if they're using from this commit forward, right?
> 
> @loadams Thanks for bringing up the backward-compatibility concern. This PR "per se" is backward-compatible, but the "quantizer" extension itself might not be supported pre-ROCm5.7 because it needs the changes to the HIP cooperative groups headers.

@jithunnair-amd - thanks for clarifying, so this won't break any of the other ops on pre-5.7, it will just not build the quantizer op.  That's much better, thanks!

**Pull request merged**